mods/simple_service.js

```diff
@@ -19,74 +18,308 @@ function SimpleService(options) {
+            if (/^[0-9]+$/.test(option.toString())) {
+                return '(res["' + catchCond + '"] === ' + option + ')';
+            } else {
+                return '(/^' + option.replace('x', '.') +
+                    '$/.test(res["' + catchCond + '"].toString())';
+            }
+        } else {
+            return '(res["' + catchCond + '"] === ' + option + ')';
+        }
+    }
+
+    var condition = [];
+    var code;
+    Object.keys(catchDefinition).forEach(function(catchCond) {
+        if (Array.isArray(catchDefinition[catchCond])) {
+            var orCondition = catchDefinition[catchCond].map(function(option) {
+                return createCondition(catchCond, option);
+            });
+            condition.push('(' + orCondition.join(' || ') + ')');
+        } else {
+            condition.push(createCondition(catchCond, catchDefinition[catchCond]));
+        }
+    });
+    code =  'return (' + condition.join(' && ') + ');';
+    /* jslint evil: true */
+    return new Function('res', code);
+}
+
+/**
+ * Checks if a step definition contains 'return' or 'return_if' stanza
+ * @param stepConf step config object
+ * @returns {boolean} true if there's 'return' or 'return_if' in the step
+ */
+function isReturnStep(stepConf) {
+    return Object.keys(stepConf).some(function(requestName) {
+        return stepConf[requestName].return || stepConf[requestName].return_if;
+    });
+}
+
+/**
+ * Validates a single step in the request chain
+ *
+ * @param {object} stepConf step configuration with optional
+ *                 'request', 'return', 'return_if' and 'catch' properties.
+ */
+function validateStep(stepConf) {
+    // Returning steps can't be parallel
+    if (isReturnStep(stepConf) && Object.keys(stepConf).length > 1) {
+        throw new Error('Invalid spec. ' +
+            'Returning requests cannot be parallel. Spec: ' + JSON.stringify(stepConf));
+    }
+
+    // Either 'request' or 'return' must be specified
+    if (!Object.keys(stepConf).every(function(requestName) {
+        return stepConf[requestName].request || stepConf[requestName].return;
+    })) {
+        throw new Error('Invalid spec. ' +
+            'Either request or return must be specified. Step: ' + JSON.stringify(stepConf));
+    }
+
+    // Only supply 'return_if' when 'request' is specified
+    if (Object.keys(stepConf).some(function(requestName) {
+        return stepConf[requestName].return_if && !stepConf[requestName].request;
+    })) {
+        throw new Error('Invalid spec. ' +
+            'return_if should have a matching request. Step: ' + JSON.stringify(stepConf));
+    }
+
+    // Only supply 'catch' when 'request' is specified
+    if (Object.keys(stepConf).some(function(requestName) {
+        return stepConf[requestName].catch && !stepConf[requestName].request;
+    })) {
+        throw new Error('Invalid spec. ' +
+            'catch should have a matching request. Step: ' + JSON.stringify(stepConf));
```

Comment:
@gwicke there's a problem with bare catch: 
 - we cant's really match stuff other then a current request result, because it's implemented like this
 - if we have a catch, the request chain return undefined, which is hard to handle.

So, it's quite hard to handle, and the `catch` impl would be way slower, so I'd better not include it until we need it.

What do you think?